src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php

```diff
@@ -22,33 +22,216 @@
+            function (Host $host, &$hostTemplateIds) use (&$extractHostTemplateIdsFromHost): void {
+                foreach ($host->getTemplates() as $hostTemplate) {
+                    if ($hostTemplate->isActivated() === false) {
+                        continue;
+                    }
+                    $hostTemplateIds[] = $hostTemplate->getId();
+                    if (!empty($hostTemplate->getTemplates())) {
+                        // The recursive call here allow you to keep the priority orders of the host templates
+                        $extractHostTemplateIdsFromHost($hostTemplate, $hostTemplateIds);
+                    }
+                }
+            };
+        $extractHostTemplateIdsFromHost($host, $hostTemplateIds);
+
+        /**
+         * First, we will search for services already associated with the host to avoid creating a new one with
+         * same service description.
+         */
+        $serviceAlreadyExists = $this->findServicesByHost($host);
+
+        /**
+         * Then, we memorize the alias of service.
+         * The service description is based on the alias of service template when it was created.
+         */
+        $serviceAliasAlreadyUsed = [];
+        foreach ($serviceAlreadyExists as $service) {
+            $serviceAliasAlreadyUsed[] = $service->getDescription();
+        }
+
+        /**
+         * Then, we will search for all service templates associated with the host templates
+         */
+        $hostTemplateServices = $this->findHostTemplateServices($hostTemplateIds);
+
+        /**
+         * Extract service templates associated to host template.
+         *
+         * **We retrieve service templates only from an enabled host template.**
+         *
+         * @param int $hostTemplateId Host template id for which we want to find the services templates
+         * @return Service[]
+         */
+        $extractServiceTemplatesByHostTemplate = function (int $hostTemplateId) use ($hostTemplateServices): array {
+            $serviceTemplates = [];
+            foreach ($hostTemplateServices as $hostTemplateService) {
+                if ($hostTemplateService->getHostTemplate()->getId() === $hostTemplateId) {
+                    // Only if the host template is activated
+                    if ($hostTemplateService->getHostTemplate()->isActivated()) {
+                        $serviceTemplates[] = $hostTemplateService->getServiceTemplate();
+                    }
+                }
+            }
+            return $serviceTemplates;
+        };
+
+        $servicesToBeCreated = [];
+
+        /**
+         * Then, we set aside the services to be created.
+         * We must not have two services with the same description (alias of the service template).
+         * The priority order is defined by the list of host templates.
+         * We only retrieve the service templates that are activated.
+         */
+        if (!empty($hostTemplateIds)) {
```

Comment:
the foreach won't loop if $hostTemplateIds is empty. IMHO, this check is useless as there's no else case